src/NewTools-Debugger/StDebuggerActionModel.class.st

```diff
@@ -351,7 +351,7 @@ StDebuggerActionModel >> stackOfSize: anInteger [
 { #category : #context }
 StDebuggerActionModel >> statusStringForContext [
 
-	^ self contextPredicate printDescription
+	^ self exception description
```

Comment:
We can't throw away context predicates like this.

For example, it handles printing for post-mortems and halts.
DoIts are also handled by context predicates. DoIts are just contexts with an `OupsNullException` and that's why the debugger doesn't display them correctly after your changes.

Plus, what you do is done in `StDebuggerErrorContextPredicate`: it prints the description of the exception only if it the context is not a DoIt.
So, if I look at the problem from the beginning, the problem is simply that an `StDebuggerErrorContextPredicate` is not created because the debugger action model doesn't consider exceptions signaled via `#signalIn:` as exceptions.
As you said, this is because of the method:
```Smalltalk
StDebuggerActionModel>>#contextSignalsAnException
  ...
  and: [  aContext selector = #signal ] ]
```
So, I would propose as a fix to:
- remove `Halt>>#description`,
- reverse `StDebuggerActionModel>>#statusStringForContext` to what it was before the changes,
- in `StDebuggerActionModel>>#contextSignalsAnException`, change the line `and: [  aContext selector = #signal ] ]` to `and: [ (#signal signalIn:) includes: aContextSelector ] ]`
- Then, in `StDebuggerErrorContextPredicate>># printDescription`, we would have to handle `Halt` just the same way as `OupsNullException`. To avoid using `kindOf`, you could define an extension method in the class Exception `isDebuggerWhiteListed` that would return `true` and that would be overriden by `OupsNullException` and `Halt` to return `false`...

Reply:
Thanks for the comments. I think what is needed most is a set of additional test cases that clarify some of the intended behavior. Since I don't have a spec for the exception handling system, I can only code against guesses about what correct behavior should look like based on examples I find in the image, some of which seem like bugs, as you mention, and some of which I could not find in order to produce, such as any code that generates a post-mortem. Could you perhaps suggest some tests that clarify some of these issues and intended behaviors?

Reply:
I think you can add a test in which a debugger action model proceeds until it reaches a method with a Halt that uses `#signal` (such as `1 halt`) and we could check that the description is the same as the one we would get with `Halt if: [true]`: "Halt in class>>method`

For post-mortems, it is really hard to get one and I can't think of an easy way to do it. Generally, it happens when we materialize a stack that had been previously fueled out in a file.
In a unit test, what you could do is create one "artificially". You could create a debugger action model on any session and store the context status string into a variable. Then, you could make the context artificially by executing `debuggerActionModel contextPredicate postMortem: true` .
Then, you could check that the context status string is the same as the one you've stored into your variable but prefixed by `[Post-mortem] `.

I don't think there are more cases than these ones and the ones you already cover in your tests. I don't know what @StevenCostiou  thinks

Reply:
Thanks. I will try to get that working next time I have time to work on this. Would it be preferable to pull request just the tests, or is it preferred to have working code along with tests?

I'm trying to get my contracts/type checking library working, but it is considerably less useful than it could be if the exceptions it throws can't describe the type error.